[Gan Ting, Nicolle] iP

nicollegann opened the pull request:
## Duke
> "The shorter way to do many things is to only do one thing at a time." – Mozart ([source](https://www.inc.com/dan-scalco/22-time-management-quotes-to-inspire-you-to-achieve-your-goals.html))

Duke frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,
1. download it from [here](https://github.com/nicollegann/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉 

And it is **FREE**!

Features:
- [X] Managing tasks (to-dos, deadlines, events)
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
##
If you're a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

nicollegann committed:
add level 1: greet, echo, exit

nicollegann committed:
add level 2: add, list

nicollegann committed:
add Task.java

nicollegann committed:
update Duke.java with Task class

nicollegann committed:
add level 3: mark, unmark

nicollegann committed:
Add Todo, Deadline and Event classes

nicollegann committed:
Add level 4: support for tracking three types of tasks (todo, deadline, event)

nicollegann committed:
Change indentation in Duke

nicollegann committed:
Update expected.txt and input.txt to semi-automate the testing of Duke

nicollegann committed:
Add level 5: handle errors due to incorrect user inputs, readjust indentation

nicollegann committed:
add level 6: support for deleting tasks from the list

nicollegann committed:
use enumeration to represent commands from user input

nicollegann committed:
Update code to follow coding standard

nicollegann committed:
Extract code into separate methods in Duke class

nicollegann committed:
Load task list data from hard disk when Duke starts up

nicollegann committed:
Save the tasks in hard disk when there are changes to task list

nicollegann committed:
Store dates as java.time.LocalDateTime in task objects

nicollegann committed:
Merge branch 'branch-Level-7'

nicollegann committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java

nicollegann committed:
Refactor the code to extract out closely related code as classes

nicollegann committed:
Organize the classes into java packages

nicollegann committed:
Add some JUnit tests to test the behavior of the code

nicollegann committed:
Change line design in Ui class

nicollegann committed:
Add JavaDoc comments

nicollegann committed:
Ensure code follows coding standard

nicollegann committed:
Add find command to find a task by searching for a keyword
Replace if-else with switch statement in Parser class

nicollegann committed:
Merge branch 'branch-A-JavaDoc'

nicollegann committed:
Merge branch 'branch-A-CodingStandard'

nicollegann committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/command/InvalidCommand.java
#	src/main/java/duke/task/Task.java

takufunkai commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    /**
+     * Returns whether this is an exit command.
```
Might be clearer to explain explicitly what exit and non-exit commands will return respectively, for example `Returns false if and only if the Command is an ExitCommand`.

takufunkai commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    /**
+     * Execute the delete command, print confirmation message and saves the updated task list to file.
```
I think it should be executes, prints and saves, and I think it was repeated in several other java-docs you wrote for `Command::execute`

nicollegann replied:
Thanks for pointing out these! Your comments have been really helpful 😄

takufunkai commented on src/main/java/duke/command/ExitCommand.java:
```diff
@@ -0,0 +1,33 @@
+ */
+public class ExitCommand extends Command {
+
+    public ExitCommand() {}
```
Might be cleaner to leave it as [default constructor](https://beginnersbook.com/2014/01/default-constructor-java-example/) instead? Only scenario I can imagine to have this line is if we want to keep our constructor `private` so that other classes will not be able to call this constructor.

takufunkai commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    /**
+     * Executes the find command to find tasks containing the keyword and print them.
+     * @param taskList The current list of tasks.
```
We're told to leave a one-line buffer between parameters and the description

takufunkai commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,144 @@
+                if (descriptionExists(inputArr, "Oops, a todo description cannot be left empty!")) {
+                    Task task = new Todo(description);
+                    return new AddCommand(task);
+                }
+
+            case "deadline":
+                if (descriptionExists(inputArr, "Oops, a deadline description cannot be left empty!")) {
+                    String[] descrArr = description.split(" /by ");
+                    if (keywordExists(descrArr, "/by", "deadline")) {
+                        LocalDateTime dateTime = parseDateTime(descrArr[1]);
+                        Task task = new Deadline(descrArr[0], dateTime);
+                        return new AddCommand(task);
+                    }
+                }
+
+            case "event":
+                if (descriptionExists(inputArr, "Oops, an event description cannot be left empty!")) {
+                    String[] descrArr = description.split(" /at ");
+                    if (keywordExists(descrArr, "/at", "event")) {
+                        LocalDateTime dateTime = parseDateTime(descrArr[1]);
+                        Task task = new Event(descrArr[0], dateTime);
+                        return new AddCommand(task);
+                    }
+                }
+
+            case "find":
+                if (descriptionExists(inputArr, "Please specify a keyword!")) {
+                    if (inputArr.length > 2) {
+                        throw new DukeException("Please specify only one keyword!");
+                    } else {
+                        return new FindCommand(description);
+                    }
+                }
+
+            case "bye":
+                return new ExitCommand();
+
+            default:
+                throw new DukeException("Oh no! I don't understand what that means...");
+            }
```
Switch-case statements should not have a line-break between them.

takufunkai commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,49 @@
+     * @throws DukeException If updated task list cannot be saved to file.
+     */
+    public void execute(TaskList taskList, Ui ui, Storage storage) throws DukeException {
+        taskList.add(this.task);
+        ui.printOutput("Got it! I've added this task:\n  "
+                + ui.showIndent() + this.task + "\n"
+                + ui.showIndent() + taskList.getListStatus());
+
+        storage.saveToHardDisk(taskList);
```
I really like that your code is segmented into logical blocks of work!

takufunkai reviewed (commented):
Your program abides by most if not all the Java coding standards, and your Java Docs are also well written. You kept all your fields private, and also segmented some of your code into logical blocks of work. Your methods also flow logically to some degree. I think you made this very easy for me 😅😅💯

nicollegann committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

nicollegann committed:
Set up gradle

nicollegann committed:
Fix coding style violations based on checkstyle

nicollegann committed:
Update javadocs to follow coding standards

nicollegann committed:
Merge branch 'branch-A-Gradle'

petermonky commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,52 @@
+                    + taskDeleted + "\n" + ui.showIndent() + taskList.getListStatus());
+            storage.saveToHardDisk(taskList);
+        } else {
+            throw new DukeException("Invalid task number. There are " + currentTotalTasks + " in the list.");
```
I like how detailed your `DukeException` messages are. I can definitely see it helping the user.

petermonky commented on src/main/java/duke/command/InvalidCommand.java:
```diff
@@ -0,0 +1,32 @@
+/**
+ * Represents an invalid command.
+ */
+public class InvalidCommand extends Command {
```
I like how you have a separate class for an invalid command; I would not have thought to handle invalid commands like this.

petermonky commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,154 @@
+ */
+public class Parser {
+
+    public Parser() {}
```
I don't think this constructor is necessary; if I'm not wrong, Java implicitly provides the default constructor for all classes.

petermonky commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,154 @@
+            throw new DukeException("Invalid date/time format.\n    "
+                    + "Please use the following format: yyyy-mm-dd HH:mm");
+        }
+        return new InvalidCommand();
```
Not sure if I'm understanding this right, but is this line ever reached? I feel like the default clause of the switch-case statement already handles invalid commands.

petermonky commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,154 @@
+     * @param descrArr The description split by the keyword.
+     * @param keyword The keyword to be found in the description.
+     * @param taskType The type of the task (e.g. deadline, event)
+     * @return True if keyword can be found, false otherwise.
```
Is this `@return` descriptor accurate? The method only seems to return `true`.

petermonky commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,154 @@
+     * Returns LocalDateTime.
+     *
+     * @param dt The string of format yyyy-MM-dd HH:mm
+     * @return LocalDateTime.
```
Perhaps could elaborate on what `LocalDateTime` it is? i.e. the parsed `LocalDateTime`?

petermonky commented on src/main/java/duke/util/Storage.java:
```diff
@@ -0,0 +1,110 @@
+            }
+        } catch (IOException e) {
+            System.out.println("Unable to create file to save tasks.");
+        }
```
I think this part can be modified to make `filePath` more accurate; as of now `filePath` is being treated more like the path of the directory that the file is located in. Perhaps could make use of `getParentFile()` and `mkdirs()`? i.e. from

```java
File folder = new File(filePath);
try {
    if (!folder.exists()) {
        folder.mkdir();
    } else {
        File file = new File(filePath + fileName);
        file.createNewFile();
    }
} catch (IOException e) {
    System.out.println("Unable to create file to save tasks.");
}
```
to
```java
File file = new File(filePath);
try {
    if (!file.exists()) {
        file.getParentFile().mkdirs();
        file.createNewFile();
    }
} catch (IOException e) {
    System.out.println("Unable to create file to save tasks.");
}
```

petermonky commented on src/main/java/duke/util/Storage.java:
```diff
@@ -0,0 +1,110 @@
+            fw.close();
+        } catch (IOException e) {
+            throw new DukeException("Unable to save changes to hard disk :(");
+        }
```
Shouldn't `fw.close()` be in the `finally` block? Since control wouldn't return to the `try` block, should an error occur before `fw.close()`.

petermonky commented on src/main/java/duke/util/TaskList.java:
```diff
@@ -0,0 +1,161 @@
+     */
+    public String findTasksWithKeyword(String keyword) {
+        String list = "Here are the matching tasks in your list:";
+        int index = 1;
```
Is `index` used in this method, other than when it is incremented in line 152?

petermonky commented on src/main/java/duke/util/TaskList.java:
```diff
@@ -0,0 +1,161 @@
+     * @param keyword The keyword used to search for tasks.
+     * @return String representation of matching tasks.
+     */
+    public String findTasksWithKeyword(String keyword) {
```
I feel like the name of this method is a bit misleading; one might assume the method returns a `TaskList`. Perhaps the method name can be changed to better reflect the return type?

petermonky commented on src/main/java/duke/util/Ui.java:
```diff
@@ -0,0 +1,85 @@
+     */
+    public String showIndent() {
+        return indent;
+    }
```
I like how you abstracted away the string constants to individual methods!

petermonky reviewed (commented):
Overall, a very nicely structured project; clear use of encapsulation here and there. Other than some minor code quality improvements, a solid piece of work.

nicollegann committed:
Create GUI for Duke

nicollegann committed:
Merge branch 'branch-Level-10'

nicollegann committed:
Modify build.gradle to update mainClassName

nicollegann committed:
Add assertions

After adding a task to the task list, program assumes that the size
of task list will always be more than 0.

Including assertions can help to define assumptions so that the
runtime can verify them.

Let's add an assertion to the add command so that we can verify that
the size of task list is always more than 0 after a task has been
added to it.

nicollegann committed:
Modify build.gradle to enable assertions

Java disables assertions by default.

Let's enable assertions in the build.gradle file so that all
assertions are verified when running the code using gradle.

nicollegann committed:
Modify code according to suggestions from pr review

nicollegann committed:
Refactor code in Parser class

nicollegann committed:
Improve code quality

There are several classes that do not follow the code quality
guidelines.

Improving code quality can enhance readability of the code.

Let's
* use guard clauses
* include default branch in switch-case statements
* use default/else branch to detect errors
* break down complicated expressions
* use constants to replace magic literals

nicollegann committed:
Merge pull request #1 from nicollegann/branch-A-Assertions

Add assertions

nicollegann committed:
Merge branch 'master' of https://github.com/nicollegann/ip

nicollegann committed:
Merge branch 'master' into branch-A-CodeQuality

nicollegann committed:
Merge pull request #2 from nicollegann/branch-A-CodeQuality

Improve code quality

nicollegann committed:
Add update command

nicollegann committed:
Merge branch 'branch-C-Update'

nicollegann committed:
Improve GUI

nicollegann committed:
Add padding for GUI DialogBox

nicollegann committed:
Change indentation for response messages

nicollegann committed:
Improve GUI

Let's
* Add placeholder in textfield
* Close program after exit command

nicollegann committed:
Remove unused variable

nicollegann committed:
Add images in GUI

nicollegann committed:
Add User Guide and representative screenshot

nicollegann committed:
Update User Guide

nicollegann committed:
Change font to Arial

nicollegann committed:
Update User Guide with link to duke.jar
